### DISCUSSION TITLE
2023-09: 我是如何构建 backon 的？

### DISCUSSION THREAD

**Discussion by giscus** [Xuanwo/blog#398] [Announcements]
# 2023-09: 我是如何构建 backon 的？

An infrastructure engineer, focused on distributed storage system

https://xuanwo.io/reports/2023-09/

**Comment by Trojanking123** [comment]
看起来不错

**Comment by noirgif** [comment]
很有意思的设计，介绍辛苦了！下面是我的一点小意见。

我第一遍看到最后，也没有明白`fetch`是什么类型。最后的链式调用，调用的是什么trait的方法。但是我现在大概懂了，以防万一确认一下：

`fetch`是`Retryable`类型，所以`fetch.retry`提供一个 `Retry`类型。

`Retry`的`when`方法设置其`retryable`成员,`notify`方法设置其`notify`成员。我假设这里使用了 fluent interface。如果说是factory pattern的话，就差最后的`build()`调用了。`BlockingRetry`我很能理解，前面都是在设置成员，最后的`call`就类似于`build`，让`Retry`开始工作。但是非同步的话，当`Retry`类实例被创建的时候就已经有任务在执行了，如果这个任务执行完成后才设置了`retryable`和`notify`会怎么样。

所以对此我的建议是在设计章节里明确`fetch`是什么、怎么创建；或者是在最开始假设我们的需求，比如这样（带了一点我的理解）：
> 当我们想做`fetch_url(url: &str)`的时候，我们用这个参数创建一个实例`fetch`，这个实例应该有如下的功能：
> 1. 执行一个任务，这个任务可能会失败，当失败的时候需要等待重试
> 2. 提供一个`Future`，如果这个任务还在运行或者重试则为 Pending，否则返回结果
> 3. 允许用户决定重试的间隔，这个间隔可以是`Iterator<Item = Duration>`
> 4. 允许用户自定义重试的条件
> 5. 允许用户设定一个函数，这个函数在每次任务结束（成功或失败）的时候都会被调用一次（也可以和4合并在一个函数里）

---

另外一点意见，是“我不喜欢它提供的API“。一看上去，好像是个人喜好的问题，但是我觉得你有你的理由。除了”简单直接，不打乱用户的阅读顺序，一眼能定位业务逻辑位置“以外，看完全文我觉得可能还有这些理由：

1. `Backoff`不够直白。用户需要查看文档才能明白这是什么样的`trait` 
> 顺便一提，Backoff 的界面好像自[Spring Framework](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/util/backoff/BackOff.html)就已经有了。
2. 用户只能等到整个任务重试完或执行完才能获得结果，没有办法获得中间结果

你怎么看这篇文章的导入部分？

**Reply by Xuanwo** [reply]
> 所以对此我的建议是在设计章节里明确`fetch`是什么、怎么创建；或者是在最开始假设我们的需求，

感觉很棒！
